platform/mbed_error_report.c

```diff
@@ -0,0 +1,278 @@
+#endif
+
+#if DEVICE_ITM && (MBED_CONF_ERROR_REPORT_INTERFACE==DEVICE_ITM)
+    /*Initialize ITM interfaces*/
+    mbed_itm_send(ITM_PORT_SWO, ch);
+#endif
+}
+
+/* Limited print functionality which prints the string out to 
+stdout/uart without using stdlib by directly calling serial-api 
+and also uses less resources 
+The fmtstr contains the format string for printing and for every %
+found in that it fetches a uint32 value from values buffer
+and prints it in hex format.
+*/
+void mbed_error_print(char *fmtstr, uint32_t *values)
+{
+#if DEVICE_SERIAL || DEVICE_ITM
+    int i = 0;
+    int idx = 0;
+    int vidx = 0;
+    char num_str[9]={0};
+    char *str=NULL;
+    
+    //Init error reporting interfaces
+    mbed_error_init();
+        
+    while(fmtstr[i] != '\0') {
+        if(fmtstr[i]=='%') {
+            i++;
+            if(fmtstr[i]=='x') {
+                //print the number in hex format
+                value_to_hex_str(values[vidx++],num_str);
+                for(idx=7; idx>=0; idx--) {
+                    mbed_error_putc(num_str[idx]);
+                }
+            }
+            else if(fmtstr[i]=='d') {
+                memset(num_str, '0', sizeof(num_str));
+                //print the number in dec format
+                value_to_dec_str(values[vidx++],num_str);
+                idx=5;//Start from 5 as we dont have big decimal numbers
+                while(num_str[idx--]=='0' && idx > 0);//Dont print zeros at front
+                for(idx++;idx>=0; idx--) {
+                    mbed_error_putc(num_str[idx]);
+                }
+            }
+            else if(fmtstr[i]=='s') {
+                //print the string
+                str = (char *)((uint32_t)values[vidx++]);
+                while(*str != '\0') {
+                    mbed_error_putc(*str);
+                    str++;
+                }
+                str = NULL;
+            } else {
+                //Do not handle any other % formatting and keep going
+            }
+        } else {
+            //handle carriage returns
+            if (fmtstr[i] == '\n') {
+                mbed_error_putc('\r');
+            }
+            mbed_error_putc(fmtstr[i]);
+        }
+        i++;
+    }
+#endif    
+}
+
+#ifdef MBED_CONF_RTOS_PRESENT 
+/* Prints thread info from a list */
+void print_threads_info(osRtxThread_t *threads)
+{
+    while(threads != NULL) {
+        print_thread( threads );
+        threads = threads->thread_next;
+    }
+}
+
+/* Prints info of a thread(using osRtxThread_t struct)*/
+void print_thread(osRtxThread_t *thread)
+{
+    uint32_t data[5];
+
+    data[0]=thread->state;
+    data[1]=thread->thread_addr;
+    data[2]=thread->stack_size;
+    data[3]=(uint32_t)thread->stack_mem;
+    data[4]=thread->sp;
+    mbed_error_print("\nState: 0x%x Entry: 0x%x Stack Size: 0x%x Mem: 0x%x SP: 0x%x", data);
+}
+#endif
+
+/* Prints the error information */
+void mbed_report_error(mbed_error_ctx *error_ctx, char *error_msg) 
+{
```

Comment:
Query: Will this function be called in case of warning/non-fatal errors as well? If yes then I would say it is lacking synchronization with existing mbed_trace library. 
I know the prints in the call are under critical section, but prints from application and mbed_trace are not in critical section. Consider a scenario where thread1 is in between printing mbed_trace log and context switches to thread_2 which has non-fatal error. It will start printing on UART console without previous log message dumped completely. Logging in this module does not guarantee line interleaving with existing libraried